src/dht_proxy_server.cpp

```diff
@@ -0,0 +1,397 @@
+                result["ipv6"] = dht_->getNodesStats(AF_INET6).toString();
+                Json::FastWriter writer;
+                s->close(restbed::OK, writer.write(result));
+            }
+            else
+                s->close(restbed::NOT_FOUND, "{\"err\":\"Incorrect DhtRunner\"}");
+        }
+    );
+}
+
+void
+DhtProxyServer::get(const std::shared_ptr<restbed::Session>& session) const
+{
+    static constexpr dht::InfoHash INVALID_ID {};
+    const auto request = session->get_request();
+    int content_length = std::stoi(request->get_header("Content-Length", "0"));
+    auto hash = request->get_path_parameter("hash");
+    session->fetch(content_length,
+        [&](const std::shared_ptr<restbed::Session> s, const restbed::Bytes& b)
+        {
+            (void)b;
+            if (dht_) {
+                InfoHash infoHash(hash);
+                if (infoHash == INVALID_ID) {
+                    infoHash = InfoHash::get(hash);
+                }
+                Json::FastWriter writer;
+                dht_->get(infoHash, [s, &writer](std::shared_ptr<Value> value) {
+                    // Send values as soon as we get them
+                    Json::Value result;
+                    s->yield(restbed::OK,  writer.write(value->toJson()));
+                    return true;
+                }, [s, &writer](bool ok) {
+                    // Communication is finished
+                    auto response = std::to_string(ok);
+                    s->close(restbed::OK, "{\"ok\": " + response + "}");
+                });
+            } else {
+                s->close(restbed::NOT_FOUND, "{\"err\":\"Incorrect DhtRunner\"}");
+            }
+        }
+    );
+}
+
+void
+DhtProxyServer::listen(const std::shared_ptr<restbed::Session>& session) const
+{
+    static constexpr dht::InfoHash INVALID_ID {};
+    const auto request = session->get_request();
+    int content_length = std::stoi(request->get_header("Content-Length", "0"));
+    auto hash = request->get_path_parameter("hash");
+    InfoHash infoHash(hash);
+    if (infoHash == INVALID_ID)
+        infoHash = InfoHash::get(hash);
+    size_t token;
+    session->fetch(content_length,
+        [&](const std::shared_ptr<restbed::Session> s, const restbed::Bytes& b)
+        {
+            (void)b;
+            if (dht_) {
+                Json::FastWriter writer;
+                token = dht_->listen(infoHash, [s, &writer](std::shared_ptr<Value> value) {
+                    // Send values as soon as we get them
+                    if (!s->is_closed())
+                        s->yield(restbed::OK,  writer.write(value->toJson()));
+                    return !s->is_closed();
+                }).get();
+                // Handle client deconnection
+                // NOTE: for now, there is no handler, so we test the session in a thread
+                listenThreads_.emplace_back(std::shared_ptr<std::thread>(
+                    new std::thread([&]() {
```

Comment:
* don't capture by reference in async operations.
* use a single common thread to check all pending operations
* use a std::set of pending Sp<restbed::Session> instead of a vector of threads
* don't forget to remove a session from the set once closed (currently threads remain in the vector forever)
* (just so you know) prefer vector<thread> instead of vector<shared_ptr<thread>>

Reply:
"use a single common thread to check all pending operations" => done
"use a std::set of pendingSp<restbed::Session> instead of a vector of threads" => I need to store token and hash, so I use a vector with a struct. The vector is sufficient here because the proxy must not have millions of connections and see NOTE below.
"don't forget to remove a session from the set once closed " => done

NOTE: restbed 5.0 have a lot of great features and when it will be released, we can remove the thread about listeners and handle client close. Do we choose to embed the same version as `ring-daemon` or the version embedded in Ubuntu or directly from the git?